include/nbl/builtin/hlsl/workgroup/arithmetic.hlsl

```diff
@@ -15,64 +16,81 @@ namespace hlsl
-struct reduction
+T reduction(T value, NBL_REF_ARG(SharedAccessor) accessor)
 {
-    struct inclusive_scan_t : subgroup::inclusive_scan<T, Binop> {}; // Yes, inclusive scan subgroup op is used for reduction workgroup ops
-    T operator()(T value)
-    {
-        SharedAccessor accessor;
-        accessor.main.workgroupExecutionAndMemoryBarrier();
-        WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor> wsh = WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor>::create(false, Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
-        T result = wsh(value);
-        accessor.main.workgroupExecutionAndMemoryBarrier();
-        T retVal = Broadcast<uint, SharedAccessor>(result, wsh.lastInvocationInLevel);
-        accessor.main.workgroupExecutionAndMemoryBarrier();
-        return retVal;
-    }
-};
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    WSHT wsh = WSHT::create(Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+    wsh(value, accessor);
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    T retVal = Broadcast<uint, SharedAccessor>(wsh.lastLevelScan, accessor, wsh.lastInvocationInLevel);
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    return retVal;
+}
 
 template<typename T, class Binop, class SharedAccessor>
-struct inclusive_scan
+T inclusive_scan(T value, NBL_REF_ARG(SharedAccessor) accessor)
 {
-    struct inclusive_scan_t : subgroup::inclusive_scan<T, Binop> {};
-    T operator()(T value)
-    {
-        SharedAccessor accessor;
-        accessor.main.workgroupExecutionAndMemoryBarrier();
-        WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor> wsh = WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor>::create(true, Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
-        wsh(value);
-        WorkgroupScanTail<T, Binop, SharedAccessor> wst = WorkgroupScanTail<T, Binop, SharedAccessor>::create(false, Binop::identity(), wsh.firstLevelScan, wsh.lastInvocation, wsh.scanStoreIndex);
-        T retVal = wst();
-        accessor.main.workgroupExecutionAndMemoryBarrier();
-        return retVal;
-    }
-};
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    WSHT wsh = WSHT::create(Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+    wsh(value, accessor);
+    WSTT(false) incl_wst = WSTT(false)::create(Binop::identity(), wsh);
+    T retVal = incl_wst(accessor);
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    return retVal;
+}
 
 template<typename T, class Binop, class SharedAccessor>
-struct exclusive_scan
+T exclusive_scan(T value, NBL_REF_ARG(SharedAccessor) accessor)
 {
-    struct inclusive_scan_t : subgroup::inclusive_scan<T, Binop> {}; // Yes, inclusive scan subgroup op is used for exclusive workgroup ops
-    T operator()(T value)
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    WSHT wsh = WSHT::create(Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+    wsh(value, accessor);
+    WSTT(true) excl_wst = WSTT(true)::create(Binop::identity(), wsh);
```

Comment:
for clarity that WSHT and WSTT share state I'd recommendly creating only the Tail, and instead of passing the head to `WSTT::create` forward the `WSHT::create`'s arguments